File: AKAuth0TestApp/AKAuth0TestApp/ViewController.m
//
//  ViewController.m
//  AKAuth0TestApp
//

#import "ViewController.h"
#import "AppConstants.h"
#import <Lock/Lock.h>
#import <SimpleKeychain.h>

static NSString *kIdTokenKeychainName = @"id_token";
static NSString *kAccessTokenKeychainName = @"access_token";
static NSString *kRefreshTokenKeychainName = @"refresh_token";
static NSString *kProfileKeychainName = @"profile";
static NSString *kKeychainName = @"Auth0";

@interface ViewController ()

- (IBAction)clickLinkedInButton:(id)sender;
- (IBAction)clickInstagramButton:(id)sender;
- (IBAction)clickTwitterButton:(id)sender;
- (IBAction)clickOpenLockUIButton:(id)sender;
- (IBAction)clickMicrosoftAccountButton:(id)sender;
- (IBAction)clickFacebookButton:(id)sender;
- (IBAction)clickGoogleButton:(id)sender;
- (IBAction)clickLogoutButton:(id)sender;

@property (weak, nonatomic) IBOutlet UILabel *userNameLabel;
@property (weak, nonatomic) IBOutlet UILabel *userIdLabel;

@end

@implementation ViewController

- (IBAction)clickLinkedInButton:(id)sender {
    [self socialAuthenticateWithName:kLinkedInConnectionName];
}

- (IBAction)clickInstagramButton:(id)sender {
    [self socialAuthenticateWithName:kInstagramConnectionName];
}

- (IBAction)clickTwitterButton:(id)sender {
    [self socialAuthenticateWithName:kTwitterConnectionName];
}

- (IBAction)clickOpenLockUIButton:(id)sender {
    A0Lock *lock = [A0Lock sharedLock];
    
    A0LockViewController *controller = [lock newLockViewController];
    controller.closable = YES;
    controller.onAuthenticationBlock = ^(A0UserProfile *profile, A0Token *token) {
        [self storeToken:token profile:profile];
        NSLog(@"User: %@", profile);
        [self dismissViewControllerAnimated:YES completion:nil];
    };
    
    [self presentViewController:controller animated:YES completion:nil];
}

- (IBAction)clickMicrosoftAccountButton:(id)sender {
    [self socialAuthenticateWithName:kWindowsLiveConnectionName];
}

- (IBAction)clickFacebookButton:(id)sender {
    [self socialAuthenticateWithName:kFacebookConnectionName];
}

- (IBAction)clickGoogleButton:(id)sender {
    [self socialAuthenticateWithName:kGoogleConnectionName];
}

- (void)socialAuthenticateWithName:(NSString *)connectionName {
    void(^success)(A0UserProfile *, A0Token *) = ^(A0UserProfile *profile, A0Token *token) {
        [self storeToken:token profile:profile];
        NSLog(@"User: %@", profile);
    };
    void(^error)(NSError *) = ^(NSError *error) {
        [self clearData];
        NSLog(@"Oops something went wrong: %@", error);
        
    };
    A0Lock *lock = [A0Lock sharedLock];
    [[lock identityProviderAuthenticator] authenticateWithConnectionName:connectionName
                                                              parameters:nil
                                                                 success:success
                                                                 failure:error];
    
}

- (IBAction)clickLogoutButton:(id)sender {
    [self logoutFromAllConnections];
    [self clearData];
}

//Saving JWT Tokens
- (void)storeToken:(A0Token *)token profile:(A0UserProfile *)profile
{
    A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:kKeychainName];
    [keychain setString:token.idToken forKey:kIdTokenKeychainName];
    [keychain setString:token.refreshToken forKey:kRefreshTokenKeychainName];
    [keychain setString:token.accessToken forKey:kAccessTokenKeychainName];
    [keychain setData:[NSKeyedArchiver archivedDataWithRootObject:profile] forKey:kProfileKeychainName];
    
    //show profile
    [self setupTextForUsername:profile.userId forUserId:profile.name];
}

- (void)storeIdToken:(NSString *)idToken
{
    A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:kKeychainName];
    [keychain setString:idToken forKey:kIdTokenKeychainName];
}

- (void)storeProfile:(A0UserProfile *)profile
{
    A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:kKeychainName];
    [keychain setData:[NSKeyedArchiver archivedDataWithRootObject:profile] forKey:kProfileKeychainName];
    
    //show profile
    [self setupTextForUsername:profile.userId forUserId:profile.name];
}

- (NSString *)retrieveIdToken
{
    A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:kKeychainName];
    NSString* token = [keychain stringForKey:kIdTokenKeychainName];
    return token;
}

- (NSString *)retrieveAccessToken
{
    A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:kKeychainName];
    NSString* token = [keychain stringForKey:kAccessTokenKeychainName];
    return token;
}

- (NSString *)retrieveRefreshToken
{
    A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:kKeychainName];
    NSString* token = [keychain stringForKey:kRefreshTokenKeychainName];
    return token;
}

- (A0UserProfile *)retrieveProfile
{
    A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:kKeychainName];
    A0UserProfile *profile = [NSKeyedUnarchiver unarchiveObjectWithData:[keychain dataForKey:kProfileKeychainName]];
    return profile;
}

- (void)clearData
{
    A0SimpleKeychain *keychain = [A0SimpleKeychain keychainWithService:kKeychainName];
    [keychain clearAll];
    
    //show profile
    [self setupTextForUsername:@"" forUserId:@""];
}

- (void)setupTextForUsername:(NSString *)username forUserId:(NSString *)userId
{
    dispatch_async(dispatch_get_main_queue(), ^{
        self.userNameLabel.text = username;
        self.userIdLabel.text = userId;
    });
}

- (void)logoutFromAllConnections {
    A0Lock *lock = [A0Lock sharedLock];
    A0APIClient *client = [lock apiClient];
    
    [client logout];
    [lock clearSessions];
}

@end

File: AKSwiftAuth0Test/AKSwiftAuth0Test/ViewController.swift
//
//  ViewController.swift
//  AKSwiftAuth0Test
//

import UIKit

let kIdTokenKeychainName = "id_token"
let kRefreshTokenKeychainName = "refresh_token"
let kProfileKeychainName = "profile"
let kKeychainName = "Auth0"

class ViewController: UIViewController {
    
    @IBOutlet weak var userNameLabel: UILabel!
    @IBOutlet weak var userIdLabel: UILabel!

    @IBAction func clickOpenLockUIButton(sender: AnyObject) {
        let controller = A0Lock.sharedLock().newLockViewController()
        controller.closable = true
        controller.onAuthenticationBlock = { (profile, token) in
            print("User: \(profile!)")
            self.dismissViewControllerAnimated(true, completion: nil)
        }
        self.presentViewController(controller, animated: true, completion: nil)

    }

    @IBAction func clickLinkedInButton(sender: AnyObject) {
        self.socialAuthenticateWithName(kLinkedInConnectionName)
    }

    @IBAction func clickInstagramButton(sender: AnyObject) {
        self.socialAuthenticateWithName(kInstagramConnectionName)
    }
    
    @IBAction func clickTwitterButton(sender: AnyObject) {
        self.socialAuthenticateWithName(kTwitterConnectionName)
    }

    @IBAction func clickMicrosoftAccountButton(sender: AnyObject) {
        self.socialAuthenticateWithName(kWindowsLiveConnectionName)
    }

    @IBAction func clickFacebookButton(sender: AnyObject) {
        self.socialAuthenticateWithName(kFacebookConnectionName)
    }
    
    @IBAction func clickGoogleButton(sender: AnyObject) {
        self.socialAuthenticateWithName(kGoogleConnectionName)
    }

    func socialAuthenticateWithName(name:String) {
        let success = { (profile: A0UserProfile, token: A0Token) in
            self.storeToken(token, profile:profile)
            print("User: \(profile)")
        }
        let failure = { (error: NSError) in
            print("Oops something went wrong: \(error)")
        }
        let lock = A0Lock.sharedLock()
        lock.identityProviderAuthenticator().authenticateWithConnectionName(name, parameters: nil, success: success, failure: failure)
    }
    
    @IBAction func clickLogoutButton(sender: AnyObject) {
        self.logoutFromAllConnections()
        self.clearData()
    }
    
    func logoutFromAllConnections() {
        let lock = A0Lock.sharedLock()
        lock.apiClient().logout()
        lock.clearSessions()
    }

    
    //Saving JWT Tokens
    func storeToken(token:A0Token?, profile:A0UserProfile?)
    {
        let keychain = A0SimpleKeychain(service: kKeychainName)
        self.storeIdToken(token?.idToken)
        self.storeProfile(profile)
        
        if let tkn = token, let refreshToken = tkn.refreshToken {
            keychain.setString(refreshToken, forKey: kRefreshTokenKeychainName)
        }
    }
    
    func storeIdToken(idToken:String?)
    {
        if let token = idToken {
            let keychain = A0SimpleKeychain(service: kKeychainName)
            keychain.setString(token, forKey: kIdTokenKeychainName)
        }
    }
    
    func storeProfile(profile:A0UserProfile?)
    {
        if let prof = profile {
            let keychain = A0SimpleKeychain(service: kKeychainName)
            keychain.setData(NSKeyedArchiver.archivedDataWithRootObject(prof), forKey: kProfileKeychainName)
            
            //show profile
            self.setupUsername(prof.name, userId: prof.userId)
        }
    }
    
    func retrieveIdToken() -> String?
    {
        let keychain = A0SimpleKeychain(service: kKeychainName)
        let token = keychain.stringForKey(kIdTokenKeychainName);
        return token;
    }
    
    func retrieveRefreshToken() -> String?
    {
        let keychain = A0SimpleKeychain(service: kKeychainName)
        let token = keychain.stringForKey(kRefreshTokenKeychainName);
        return token;
    }
    
    func retrieveProfile() -> A0UserProfile?
    {
        let keychain = A0SimpleKeychain(service: kKeychainName)
        let data = keychain.dataForKey(kProfileKeychainName)
        let profile = NSKeyedUnarchiver.unarchiveObjectWithData(data!)
        return profile as? A0UserProfile;
    }
    
    func clearData()
    {
        let keychain = A0SimpleKeychain(service: kKeychainName)
        keychain.clearAll()
        
        //show profile
        self.setupUsername("", userId:"")
    }
    
    func setupUsername(username: String, userId: String) {
        dispatch_async(dispatch_get_main_queue()) {
            self.userNameLabel.text = username
            self.userIdLabel.text = userId
        }
    }
    
    func showMessage(message: String) {
        dispatch_async(dispatch_get_main_queue()) {
            let alert = UIAlertController(title: "Auth0", message: message, preferredStyle: UIAlertControllerStyle.Alert)
            alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.Default, handler: nil))
            self.presentViewController(alert, animated: true, completion: nil)
        }
    }
}

